File: 0009-palindrome-number/0009-palindrome-number.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        return self.reverse_num(x)

    def reverse_num(self, x: int) -> bool:
        # Time Complexity: Number of digits in x = O(log10(x))
        # Space Complexity: O(1)
        if x < 0:
            return False
        temp = x
        x_rev = 0
        while temp:
            digit = temp % 10
            x_rev = x_rev*10 + digit
            temp = temp // 10
        if x == x_rev:
            return True
        else:
            return False

    def str_approach(self, x: int) -> bool:
        # Time Complexity: Number of digits in x = O(log10(x))
        # Space Complexity: O(1)
        if x < 0:
            return False
        x = str(x)
        return x == x[::-1]

File: 0009-palindrome-number/0009-palindrome-number.java
class Solution {
    public boolean isPalindrome(int x) {
        if ((x < 0) || (x % 10 == 0 && x != 0)) {
            // Handle negative numbers or numbers ending in 0 (except 0 itself)
            return false;
        }
        return strApproach(x);
    }

    public boolean reverseNum(int x) {
        // Time Complexity: Number of digits in x = O(log10(x))
        // Space Complexity: O(1)
        int temp = x;
        int x_rev = 0;
        while (temp > 0) {
            int digit = temp % 10;
            x_rev = (x_rev * 10) + digit;
            temp = (int) (temp / 10);
        }
        if (x == x_rev) {
            return true;
        }
        else {
            return false;
        }
    }

    public boolean strApproach(int x) {
        // Time Complexity: Number of digits in x = O(log10(x))
        // Space Complexity: O(1)
        String xstr = Integer.toString(x);
        String xstr_rev = new StringBuffer(xstr).reverse().toString();
        return xstr.equals(xstr_rev);
    }
}